### ISSUE TITLE
Error while saving a 14M point cloud to asc file 

### ISSUE THREAD

**Issue by harrandt** [CloudCompare/CloudCompare#1963]
We came across a strange problem and have no clue on what to do.
While saving a certain point cloud with 14M points to a asc file the operation is aborted with

```
[I/O] CC has caught an unhandled exception while saving file 'Cloud_nosf.txt'
An error occurred while saving 'Cloud_nosf.txt': see console
```

It happens on two different laptops with the CC versions 2.12.4, 2.13.alpha and 2.13.beta.
OS is one with Windows 10, the other 11, both have a NVIDIA RTX A2000 Laptop GPU.

The point cloud is loaded from a bin file, it happens with or without a scalar field.

I created a different point cloud with just random numbers and 14M points 
to see if there are RAM issues, etc. But that cloud exports like a charm.

Any idea on how to further debug this. Unfortunately I can't upload the problematic data as
it contains proprietary data of a customer.

Any ideas on what might cause the problem? Is there further debug info that I can provide, except the point cloud?
I will try to run some obfuscating operation on it and see if the issue persists and I can provide it.

**Comment by dgirardeau** [comment]
Can you reproduce the issue with only a few points? (using the scissors tool to keep only a very small portion).

If yes, can you send me these few points to admin@cloudcompare.org?

**Comment by harrandt** [comment]
I was trying to export a cross section but got this (2.13.alpha) error and CC crashed afterwards.

![grafik](https://github.com/CloudCompare/CloudCompare/assets/116627373/5839180c-1f34-43ef-8bba-1a2b401d2dd1)

After a reboot I saw that this also happens with 2.13.beta.

For Segment it happens, too. With polygon and rectangular and with segment in and out. The Segment in/out buttons work and shows the correct segment. But only once you push the green tick (Enter/Confirm) it crashes with the same message.

For Cross Section, placing the xyz markers works and shows the correct sections. It crashes when you press the Slice buttons (Export selection as new entity or Export multiple entities).

I also tried different sizes for segments and sections but could not figure out a pattern, all crash.

**Comment by dgirardeau** [comment]
That's very weird 😅 . Does it happen with any cloud or just with this very specific cloud? (in which case we'll really need to access this cloud!). I tried yesterday with different clouds from 8M. to 44M. and dozens of scalar fields, but I couldn't reproduce your issue(s).

**Comment by harrandt** [comment]
Yes, it's also werid on my side :) We use CC on a daily basis and only that pc was acting up. I will try to read it again from the ASCII source. The latest tests were made with a bin file.
My first though was that it is too large so I created a slightly larger cloud with random points which behaved like a charm.

**Comment by harrandt** [comment]
I tried to reproduce the steps to get the problematic point cloud but without success.
Originally I was importing two 7M point clouds from asc, then I did some transformations to overlap them, afterwards a ICP fine registration, then I merged them to form a 14M cloud. 
No removal of duplicate or very close points was performed.

Saving it as bin works, saving it as asc fails.
I will try to obfuscate the problematic file, maybe I can provide it then.

**Comment by dgirardeau** [comment]
Do you still reproduce the issue if you multiply the X and Y coordinates by 0? (with Edit > Multiply/Scale'). I guess it should obfuscate things enough...

**Comment by harrandt** [comment]
Good idea! Yes, indeed I can reproduce the error, I will upload the cloud to our fileserver and send an email with the link to the email you mentioned.

**Comment by harrandt** [comment]
When exporting to asc the file is partially written (9M of 14M pts).

**Comment by dgirardeau** [comment]
Anyway I would prefer that you share the BIN file is possible. So that I can look at it (it's probably the source of the issue, it must be somehow corrupted). At least if when loading this BIN file I can reproduce the issues you had.

**Comment by harrandt** [comment]
Of course I shared the bin file :)

**Comment by harrandt** [comment]
Did you get it?

**Comment by dgirardeau** [comment]
Yes, and I managed to reproduce the issue! Just need to debug now...

**Comment by dgirardeau** [comment]
So it appears that this cloud has only 9M. colors for 14M. points!

I really hope it's either a corrupted BIN file, or an old version of CC that had generated this file 😅 

If you want to save the file, but lose the color, you can simply remove the colors (Edit > Colors > Clear'. But it won't be possible to retrieve the missing ones...

**Comment by harrandt** [comment]
It was generated with 2.12.4, the latest stable.
I assigned a unique color to one of the two 9M pc, the other one was default blue. Then I translated and merged them.

**Comment by harrandt** [comment]
I can confirm your suggested workaround, it works when I remove the colors

**Comment by dgirardeau** [comment]
Digging in the past bug fixes, I found this one that was fixed in the 2.13.alpha version :
* Bug fix: setting the cloud RGB color could create an ill-size RGB table, which would in turn make some operations fail (such as merging clouds) (be1a87d1b0679522ed726cf0c0e79e9e89d48248)

This is probably the source of your issue. I totally forgot about it...

**Comment by harrandt** [comment]
Yes, that totally sounds like it, should I close this one?

**Comment by dgirardeau** [comment]
Yes that's part of the 2.13 official version (and the 2.13.alpha and beta versions). But any cloud merged with the 2.12 version could have this issue, and therefore would still be corrupted...

We can close this issue.
